akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterSpecKit.scala

```diff
@@ -253,4 +259,133 @@ trait GraphInterpreterSpecKit extends AkkaSpec {
       .init()
   }
 
+  abstract class OneBoundedSetup[T](ops: Array[GraphStageWithMaterializedValue[_, _]]) extends Builder {
+
+    def this(ops: Iterable[Stage[_, _]]) = {
+      this(ops.map { op ⇒
+        new PushPullGraphStage[Any, Any, Any](
+          (_) ⇒ op.asInstanceOf[Stage[Any, Any]],
+          Attributes.none)
+      }.toArray.asInstanceOf[Array[GraphStageWithMaterializedValue[_, _]]])
+    }
+
+    val upstream = new UpstreamOneBoundedProbe[T]
+    val downstream = new DownstreamOneBoundedPortProbe[T]
+    var lastEvent = Set.empty[TestEvent]
+
+    sealed trait TestEvent
+
+    case object OnComplete extends TestEvent
+    case object Cancel extends TestEvent
+    case class OnError(cause: Throwable) extends TestEvent
+    case class OnNext(elem: Any) extends TestEvent
+    case object RequestOne extends TestEvent
+    case object RequestAnother extends TestEvent
+
+    private def run() = interpreter.execute(Int.MaxValue)
+
+    private def initialize(): Unit = {
+      import GraphInterpreter.Boundary
+
+      var i = 0
+      val attributes = Array.fill[Attributes](ops.length)(Attributes.none)
+      val ins = Array.ofDim[Inlet[_]](ops.length + 1)
+      val inOwners = Array.ofDim[Int](ops.length + 1)
+      val outs = Array.ofDim[Outlet[_]](ops.length + 1)
+      val outOwners = Array.ofDim[Int](ops.length + 1)
+
+      ins(ops.length) = null
+      inOwners(ops.length) = Boundary
+      outs(0) = null
+      outOwners(0) = Boundary
+
+      while (i < ops.length) {
+        val stage = ops(i).asInstanceOf[PushPullGraphStage[_, _, _]]
+        ins(i) = stage.shape.inlet
+        inOwners(i) = i
+        outs(i + 1) = stage.shape.outlet
+        outOwners(i + 1) = i
+        i += 1
+      }
+
+      manualInit(GraphAssembly(ops, attributes, ins, inOwners, outs, outOwners))
+      interpreter.attachUpstreamBoundary(0, upstream)
+      interpreter.attachDownstreamBoundary(ops.length, downstream)
+
+      interpreter.init()
+
+    }
+
+    initialize()
+    run() // Detached stages need the prefetch
+
+    def lastEvents(): Set[TestEvent] = {
+      val events = lastEvent
+      lastEvent = Set.empty
+      events
+    }
+
+    class UpstreamOneBoundedProbe[T] extends UpstreamBoundaryStageLogic[T] {
+      val out = Outlet[T]("out")
+      out.id = 0
```

Comment:
if there would be multiple outputs. I think this can be pulled up to the parent class though.